File: src/Level/grid_square.py
from Level.Clue.clue import Clue

class GridSquare:
    """ Represents a square in the Minefield Grid """
    
    def __init__(self, row, column):
        """ Initialize the Grid Square """
        self.row = row
        self.column = column
        self.scanned = False
        self.clue = Clue()
        
        self.groundDefense = None
        
    def setGroundDefense(self, defenseObject):
        """ Set the Ground Defense """
        self.groundDefense = defenseObject
        
    def onMove(self, drone):
        """ Perform the On Move Event """
        if self.hasGroundDefense():
            self.groundDefense.onMove(drone)
        
    def scan(self, drone):
        """ Scan the Grid Square """
        self.scanned = True
        if self.hasGroundDefense():
            self.groundDefense.scan(drone)
        self.clue.populate(drone.minefield, self.row, self.column)
        
    def defuse(self, drone):
        """ Defuse the Grid Square """
        if self.hasGroundDefense():
            self.groundDefense.defuse(drone)
            
    def defuseCarefully(self, drone):
        """ Defuse the Grid Square """
        if self.hasGroundDefense():
            self.groundDefense.defuseCarefully(drone)
            
    def hitByEMP(self, drone):
        """ The Square is hit by an EMP """
        if self.hasGroundDefense():
            self.groundDefense.hitByEMP(drone)
            
    def getGroundDefense(self):
        if self.hasGroundDefense():
            return self.groundDefense
            
    def hasGroundDefense(self):
        """ Return if the Grid Square has Ground Defense """
        return self.groundDefense is not None
            
    def mined(self):
        """ Return if the Square has a Mine """
        if self.hasGroundDefense():
            return self.groundDefense.isMine()
        return False
        
    def hasWeaponry(self):
        """ Return if the Square has Weaponry """
        if self.hasGroundDefense():
            return self.groundDefense.isWeapon()
        return False
        
    def reversed(self):
        """ Return if the Grid Square causes readings to be reversed """
        if self.hasGroundDefense():
            return self.groundDefense.reverseReadings
        return False
        
    def fragile(self):
        """ Returns if the square's defenses are fragile """
        if self.hasGroundDefense():
            return self.groundDefense.fragile
        return False
        
    def __repr__(self):
        """ Print the square as a string """
        if self.scanned:
            return str(self.clue)
        else:
            return "#"

File: src/android/src/com/kaottessur/pymine/level/GridSquare.java
package com.kaottessur.pymine.level;

import com.kaottessur.pymine.GridLocationInterface;
import com.kaottessur.pymine.defense.DefenseInterface;
import com.kaottessur.pymine.level.clue.Clue;

public class GridSquare implements GridLocationInterface {
	private int row;
	private int column;
	private boolean scanned = false;
	
	private DefenseInterface defense;
	private Clue clue;
	
	public GridSquare(int row, int column) {
		this.row = row;
		this.column = column;
		clue = new Clue();
	}
	
	public int getRow() {
		return row;
	}
	
	public int getColumn() {
		return column;
	}
	
	public Clue getClue() {
		return clue;
	}
	
	public boolean isScanned() {
		return scanned;
	}
	
	public void scan(Drone drone) {
		scanned = true;
		clue.populate(drone.getMinefield(), drone.getRow(), drone.getColumn());
		if (hasDefense())
			defense.scan(drone);
	}
	
	public void defuse(Drone drone) {
		if (hasDefense())
			defense.defuse(drone);
	}
	
	public void defuseCarefully(Drone drone) {
		if (hasDefense())
			defense.defuseCarefully(drone);
	}
	
	public void performEMP(Drone drone) {
		if (hasDefense())
			defense.hitByEMP(drone);
	}
	
	public boolean hasDefense() {
		return defense != null;
	}
	
	public DefenseInterface getDefense() {
		return defense;
	}
	
	public void setDefense(DefenseInterface defense) {
		this.defense = defense;
	}

	public boolean mined() {
		if (hasDefense())
			return defense.isMine();
		return false;
	}
	
	public boolean reversed() {
		if (hasDefense())
			return defense.reverseReadings();
		return false;
	}
	
	public boolean isFragile() {
		if (hasDefense())
			return defense.isFragile();
		return false;
	}
}
